pkg/zone/feature/onoff/group.go

```diff
@@ -0,0 +1,210 @@
+			return g.client.UpdateOnOff(ctx, request)
+		}
+	}
+	allRes, allErrs := run.Collect(ctx, run.DefaultConcurrency, fns...)
+
+	err := multierr.Combine(allErrs...)
+	if len(multierr.Errors(err)) == len(g.names) {
+		return nil, err
+	}
+
+	if err != nil {
+		if g.logger != nil {
+			g.logger.Warn("some hvacs failed to get", zap.Errors("errors", multierr.Errors(err)))
+		}
+	}
+	return mergeOnOff(allRes)
+}
+
+func (g *Group) PullOnOff(request *traits.PullOnOffRequest, server grpc.ServerStreamingServer[traits.PullOnOffResponse]) error {
+	if len(g.names) == 0 {
+		return status.Error(codes.FailedPrecondition, "zone has no on off names")
+	}
+
+	type c struct {
+		name string
+		val  *traits.OnOff
+	}
+	changes := make(chan c)
+	defer close(changes)
+
+	group, ctx := errgroup.WithContext(server.Context())
+	for _, name := range g.names {
+		name := name
+		group.Go(func() error {
+			return pull.Changes(ctx, pull.NewFetcher(
+				func(ctx context.Context, changes chan<- c) error {
+					stream, err := g.client.PullOnOff(ctx, &traits.PullOnOffRequest{Name: name})
+					if err != nil {
+						return err
+					}
+					for {
+						res, err := stream.Recv()
+						if err != nil {
+							return err
+						}
+						for _, change := range res.Changes {
+							err := chans.SendContext(ctx, changes, c{name: name, val: change.OnOff})
+							if err != nil {
+								return err
+							}
+						}
+					}
+				},
+				func(ctx context.Context, changes chan<- c) error {
+					res, err := g.client.GetOnOff(ctx, &traits.GetOnOffRequest{Name: name})
+					if err != nil {
+						return err
+					}
+					return chans.SendContext(ctx, changes, c{name: name, val: res})
+				},
+			), changes)
+		})
+	}
+
+	group.Go(func() error {
+		all := make(map[string]*traits.OnOff, len(g.names))
+
+		var last *traits.OnOff
+		eq := cmp.Equal()
+		filter := masks.NewResponseFilter(masks.WithFieldMask(request.ReadMask))
+
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case change := <-changes:
+				all[change.name] = change.val
+
+				values, err := mergeOnOffMap(all, change.val)
+				filter.Filter(values)
+				if eq(last, values) {
+					continue
+				}
+				last = values
+				err = server.Send(&traits.PullOnOffResponse{Changes: []*traits.PullOnOffResponse_Change{
+					{
+						Name:       request.Name,
+						ChangeTime: timestamppb.Now(),
+						OnOff:      values,
+					},
+				}})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	})
+
+	return group.Wait()
+}
+
+func mergeOnOff(all []*traits.OnOff) (*traits.OnOff, error) {
+	switch len(all) {
+	case 0:
+		return nil, status.Errorf(codes.FailedPrecondition, "no onoff devices in group")
+	case 1:
+		return all[0], nil
+	default:
+		state := all[0].State
+		var err error
+		// either they are all off, all on or it is an error
+		// this should really return 'Mixed' or something I think, but we don't have that in the trait
+		for _, v := range all[1:] {
+			if v == nil || v.State != state {
```

Comment:
I think you should be able to ignore unspecified responses here, they shouldn't trigger the error.